sycl/include/CL/sycl/detail/pi.h

```diff
@@ -1103,7 +1110,8 @@ __SYCL_EXPORT pi_result piclProgramCreateWithSource(pi_context context,
 __SYCL_EXPORT pi_result piProgramCreateWithBinary(
     pi_context context, pi_uint32 num_devices, const pi_device *device_list,
     const size_t *lengths, const unsigned char **binaries,
-    pi_int32 *binary_status, pi_program *ret_program);
+    pi_int32 *binary_status, const pi_device_binary_property *metadata,
```

Comment:
@steffenlarsen : have you followed up this investigation?

>> Here, could we change the interface to take existing pi_device_binary instead of just native binary, and add metadata to that structure instead?

> Good question. I will investigate, but at first glance in sycl/source/detail/program_manager/program_manager.cpp#L427 it looks to me like the binary can be separate from the pi_device_binary object, so I imagine it is limited how many parameters it can replace.

And you certainly need to add function comments about the new/changed arguments.

Reply:
Apologies! I have investigated a bit further, but I have to conclude that using `pi_device_binary` directly won't reduce the number of parameters as is, as in some paths the binary isn't attached to it and is instead loaded from a file. I believe it would require some refactoring of the type which I think would be outside the scope of this patch. That said, I have noticed that the new parameters should be before the `binary_status` parameter so I will move these when I add the comments.

> And you certainly need to add function comments about the new/changed arguments.

Just to make sure we are on the same page, you are referring to documentation comments akin to the ones currently there for most `piext*` functions?

Reply:
> using pi_device_binary directly won't reduce the number of parameters as is, as in some paths the binary isn't attached to it and is instead loaded from a file

Isn't it "put" into `pi_device_binary` after loaded from a file? 

> it would require some refactoring of the type which I think would be outside the scope of this patch

I agree it would better go in a separate change-set, but want to agree on the path-forward beforehand.

>  documentation comments akin to the ones currently there for most piext* functions?

Yes, comments at function declaration please.

Reply:
> Isn't it "put" into `pi_device_binary` after loaded from a file?

Inside `program_manager.cpp`, `piProgramCreateWithBinary` is used in `createBinaryProgram`. This function is in turn used in two paths, one where the binary is contained in `pi_device_binary` and one where it is read from file using `PersistentDeviceCodeCache::getItemFromDisc`. It is true that there seem to be some caching involved, but looking at the function for reading the binary from file I do not see it being moved into `pi_device_binary` at any point before being passed to `createBinaryProgram`. I do not know the entire motivation for the two paths, so it may be possible to just store the binary data inside `pi_device_binary`, though I fear for the possible repercussions.

> I agree it would better go in a separate change-set, but want to agree on the path-forward beforehand.

I think the previous discussion with @AlexeySachkov (https://github.com/intel/llvm/pull/3735#discussion_r643428070) may be somewhat relevant to this.

Reply:
Thanks for investigation @steffenlarsen !

> I do not know the entire motivation for the two paths, so it may be possible to just store the binary data inside pi_device_binary, though I fear for the possible repercussions.

@romanovvlad : any insight on this please?

Reply:
@smaslov-intel @romanovvlad Should we make a separate issue for this?

Reply:
@vladimirlaz  Could you please comment?

Reply:
@smaslov-intel @romanovvlad @vladimirlaz 

This discussion has been dormant for a couple of weeks now. Would you mind if we go ahead with these changes and move this discussion to a separate issue and return to it when time permits?

Reply:
@steffenlarsen, sorry,  I missed this thread. Let's go ahead.

Reply:
I am OK with proceeding as is.